lib/renderers/DefaultRenderer.js

```diff
@@ -68,38 +67,72 @@ class DefaultRenderer {
     return this.renderMessage(msg, alexaEvent);
   }
 
-  renderMessage(msg, alexaEvent) {
-    const toRender = ['ask', 'tell', 'say', 'reprompt', 'card.title', 'card.content', 'card.text', 'card.image.smallImageUrl', 'card.image.largeImageUrl'];
-
-    return Promise.all(_(toRender)
-      .map((key) => {
-        let statement = _.at(msg, key)[0];
-        if (!statement) return null;
-        if (_.isArray(statement)) statement = _.sample(statement);
-        return this.renderStatement(statement, alexaEvent)
-          .then(rendered => _.set(msg, key, rendered));
-      })
-      .compact()
-      .value())
-      .then(() => msg);
-  }
+  renderMessage(msg, event) {
+    function deepSearchStatement(statement, alexaEvent) {
+      if (_.isObject(statement) && !_.isArray(statement)) {
+        return Promise.try(() => {
+          const objPromises = _.chain(statement)
+          .toPairs()
+          .map((item) => {
+            const key = item[0];
+            const value = item[1];
+            const isAnOpenResponse = _.includes(['ask', 'tell', 'say', 'reprompt'], key);
+            if (isAnOpenResponse && _.isArray(value)) {
+              return [key, deepSearchStatement.call(this, _.sample(value), alexaEvent)];
+            }
+            return [key, deepSearchStatement.call(this, value, alexaEvent)];
+          })
+          .flattenDeep()
+          .value();
+
+          return Promise.all(objPromises)
+          .then(result => _.chain(result).chunk(2).fromPairs().value());
+        });
+      }
 
-  renderStatement(statement, alexaEvent) {
-    const tokens = template.tokens(statement);
-    const qVariables = tokens.map((token) => {
-      const variable = this.config.variables[token];
-      if (!variable) return Promise.reject(new Error(`No such variable ${token}`));
-      return Promise.try(() => variable(alexaEvent.model, alexaEvent));
-    });
+      if (_.isString(statement)) {
+        return Promise.try(() => this.renderStatement(statement, alexaEvent));
+      }
 
-    const qAll = Promise.all(qVariables);
+      if (_.isArray(statement)) {
+        return Promise
+        .map(statement, statementItem => deepSearchStatement.call(this, statementItem, alexaEvent));
+      }
 
-    return qAll.then((vars) => {
-      const tokensWithVars = _.zip(tokens, vars);
-      const data = _.fromPairs(tokensWithVars);
+      return Promise.resolve(statement);
+    }
 
-      return template(statement, data);
-    });
+    return deepSearchStatement.call(this, msg, event);
+  }
+
+  renderStatement(statement, alexaEvent) {
```

Comment:
Yeah. It's time to start applying software engineering principles to this codebase to make things clearer. We've got too many if's based on data type, and too much special knowlege w/o names.

Reply:
@rmberrios what changes did you make regarding this comment?

Reply:
I didn't apply any change about this comment. We need to define which principles we want to stick to.

Reply:
it's related to this other, https://github.com/mediarain/voxa/pull/67#discussion_r145260960 what are the "better names"?